src/Swarm/Game/Step.hs

```diff
@@ -122,24 +119,114 @@ gameTick = do
+          | wakeUpTime - 2 <= time -> return ()
+          | otherwise -> sleepUntil rn wakeUpTime
+        Nothing ->
+          unless (isActive rob) (sleepForever rn)
+
+runRobotIDs :: (Has (State GameState) sig m, Has (Lift IO) sig m) => IS.IntSet -> m ()
+runRobotIDs robotNames = forM_ (IS.toList robotNames) $ \rn -> do
+  mr <- uses robotMap (IM.lookup rn)
+  forM_ mr (stepOneRobot rn)
+ where
+  stepOneRobot rn rob = tickRobot rob >>= insertBackRobot rn
+
+singleStep :: (Has (State GameState) sig m, Has (Lift IO) sig m) => SingleStep -> RID -> IS.IntSet -> m Bool
+singleStep ss focRID robotSet = do
+  let (preFoc, focusedActive, postFoc) = IS.splitMember focRID robotSet
+  case ss of
+    ----------------------------------------------------------------------------
+    -- run robots from the beginning until focused robot
+    SBefore -> do
+      runRobotIDs preFoc
+      gameStep .= RobotStep (SSingle focRID)
+      -- also set ticks of focused robot
+      mOldR <- use $ robotMap . at focRID
+      steps <- use robotStepsPerTick
+      forM_ mOldR $ \r -> robotMap %= IM.insert focRID (r & tickSteps .~ steps)
+      return False
+    ----------------------------------------------------------------------------
+    -- run single step of the focused robot (may skip if inactive)
+    SSingle rid | not focusedActive -> do
+      singleStep (SAfter rid) rid postFoc
+    SSingle rid | rid == focRID -> do
+      mOldR <- uses robotMap (IM.lookup focRID)
+      case mOldR of
+        Nothing -> do
+          debugLog "The debugged robot does not exist! Exiting single step mode."
+          runRobotIDs postFoc
+          gameStep .= WorldTick
+          ticks += 1
+          return True
+        Just oldR -> do
+          newR <- stepRobot oldR
+          insertBackRobot focRID newR
+          when (newR ^. tickSteps == 0) $ gameStep .= RobotStep (SAfter focRID)
+          return False
+    SSingle rid | otherwise -> do
+      mOldR <- uses robotMap (IM.lookup focRID)
+      case mOldR of
+        Nothing -> do
+          debugLog "The previously debugged robot does not exist!"
+          singleStep (SAfter rid) focRID postFoc
+        Just oldR -> do
+          newR <- tickRobotRec oldR
+          insertBackRobot focRID newR
+          when (newR ^. tickSteps == 0) $ gameStep .= RobotStep (SAfter focRID)
+          singleStep ss rid postFoc
+    ----------------------------------------------------------------------------
+    -- run robots after the focused robot (go to single step if new robot is focused)
+    SAfter rid | rid >= focRID -> do
+      runRobotIDs postFoc
+      gameStep .= RobotStep SBefore
+      ticks += 1
+      return True
+    SAfter rid | otherwise -> do
+      let postRID = snd . IS.split rid $ robotSet
+      singleStep SBefore rid postRID
+ where
+  h = hypotheticalRobot (Out VUnit emptyStore []) 0
+  debugLog txt = do
+    m <- evalState @Robot h $ createLogEntry (ErrorTrace Debug) txt
+    emitMessage m
```

Comment:
OK, I have carefully read through `singleStep` now.  I *think* it makes sense, but it is still very complicated and I am afraid it will end up being the source of many bugs with debugging mode.  I think we should still add a lot more inline comments.

Reply:
Bugs in debugging function could be annoying but at least will not bother players. 😄

I think we could unit test this to make sure it does the right thing.

Another idea I had was to run the entire integration test suite in single step mode which should produce the same result with raised timeouts.

Reply:
> Bugs in debugging function could be annoying but at least will not bother players.

Why not?  Players can use the debugger too. This is a programming game after all. :smile: 



> Another idea I had was to run the entire integration test suite in single step mode which should produce the same result with raised timeouts.

That sounds like a great idea!  Though probably we shouldn't do that as part of the normal CI test suite since it might be very slow.  But if the test suite passes in single-step mode then that would be a pretty good signal it is correct.  (Although it probably wouldn't exercise corner cases like e.g. running the `view` command so the focused robot changes in the middle of single-stepping.)

Reply:
@byorgey actually this was easy:

```diff
                   & baseRobot . robotContext . defReqs <>~ reqCtx
                   & baseRobot . machine .~ initMachine sol Ctx.empty emptyStore
+                  & gameStep .~ RobotStep SBefore
-          m <- timeout (time s) (snd <$> runStateT playUntilWin gs')
+          m <- timeout (-1) (snd <$> runStateT playUntilWin gs')
           case m of
```

On CI the normal tests take 52s and for me, they took 116.09s in single-step mode and all passed. :heavy_check_mark: